power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -720,28 +702,41 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
         };
     }
 
-    static auto scenario_update_restore(
-        MainModelImpl& model, ConstDataset const& update_data,
-        std::array<std::reference_wrapper<std::vector<Idx2D> const>, n_types> const& scenario_sequence,
-        SequenceIdx& current_scenario_sequence_cache, ComponentFlags const& is_independent,
-        std::vector<CalculationInfo>& infos) noexcept {
+    static auto scenario_update_restore(MainModelImpl& model, ConstDataset const& update_data,
+                                        ComponentFlags const& is_independent, SequenceIdx const& all_scenario_sequence,
+                                        SequenceIdx& current_scenario_sequence_cache,
+                                        std::vector<CalculationInfo>& infos) noexcept {
         auto do_update_cache = [&is_independent] {
             ComponentFlags result;
             std::ranges::transform(is_independent, result.begin(), std::logical_not<>{});
             return result;
         }();
 
+        auto const scenario_sequence = [&all_scenario_sequence, &current_scenario_sequence_cache,
+                                        &is_independent]() -> SequenceIdxView {
+            return run_functor_with_all_types_return_array(
+                [&all_scenario_sequence, &current_scenario_sequence_cache, &is_independent]<typename CT>() {
+                    constexpr auto comp_idx = index_of_component<CT>;
+                    if (std::get<comp_idx>(is_independent)) {
+                        return std::span<Idx2D const>{std::get<comp_idx>(all_scenario_sequence)};
+                    }
+                    return std::span<Idx2D const>{std::get<comp_idx>(current_scenario_sequence_cache)};
+                });
+        };
+
```

Comment:
Curiousity: Why was this moved here from sub_batch_calculation? Just to reduce complexity on the later? Besides that, the "new" logic is clear.

Reply:
> Just to reduce complexity on the latter?

yes and also to combine all the logic concerning which cache to use when into one single location.